src/workerd/jsg/value.h

```diff
@@ -468,18 +469,26 @@ class StringWrapper {
       kj::Maybe<v8::Local<v8::Object>> parentObject) {
     // TODO(cleanup): Move to a HeaderStringWrapper in the api directory.
     v8::Local<v8::String> str = check(handle->ToString(context));
-    auto result =
-        ByteString(KJ_ASSERT_NONNULL(tryUnwrap(context, str, (kj::String*)nullptr, parentObject)));
-    if (str->ContainsOnlyOneByte()) {
-      auto buf = kj::heapArray<kj::byte>(str->Length() + 1);
-      str->WriteOneByte(context->GetIsolate(), buf.begin());
-      for (auto b: buf) {
-        if (b >= 128) {
-          result.warning = ByteString::Warning::CONTAINS_EXTENDED_ASCII;
-          break;
-        }
+    auto result = ByteString(KJ_ASSERT_NONNULL(
+        tryUnwrap(context, str, static_cast<kj::String*>(nullptr), parentObject)));
+
+    v8::String::ValueView str_view(context->GetIsolate(), str);
+
+    if (str_view.is_one_byte()) [[likely]] {
+      // Storage is one-byte.
+      if (!simdutf::validate_ascii(
+              reinterpret_cast<const char*>(str_view.data8()), str_view.length())) [[unlikely]] {
+        result.warning = ByteString::Warning::CONTAINS_EXTENDED_ASCII;
+      }
+    } else if (str->ContainsOnlyOneByte()) {
+      // Storage is two-byte, but it contains only one byte.
+      KJ_ASSERT(!str_view.is_one_byte());
+      auto buf = reinterpret_cast<const char*>(str_view.data16());
+      if (!simdutf::validate_ascii(buf, str_view.length())) [[unlikely]] {
```

Comment:
Not that familiar with text encodings beyond ASCII – I assume that if ContainsOnlyOneByte() applies the most significant byte of a two-byte value is expected to be zero?

Reply:
Yes, you are correct. ASCII uses only one byte per character in the range of 0 to 127. In two-byte systems like utf-16, ascii characters are often represented with a leading zero byte followed by the ASCII value in the second byte.If ContainsOnlyOneByte returns true, it means that the value can be represented in a single byte, even though the underlying storage is two-bytes.